### PR TITLE
include cleanup of cDrawManager and all drawers

### PR DESCRIPTION
**Pull request by pbomta**
also construct draw manager member objects with the draw manager, instead of on the heap somewhere.

### CODE REVIEW

File: drawers/cMouseDrawer.cpp [hunk 1]
```diff
@@ -110,14 +113,13 @@ void cMouseDrawer::drawToolTip() {
 		int structureType = theStructure->getType();
 		if (structureType == WINDTRAP) {
 			cWindTrap * windTrap = dynamic_cast<cWindTrap*>(theStructure);
+            assert(windTrap);
```
Comment (stefanhendriks) [inline drawers/cMouseDrawer.cpp:116]: do we need this assert? (was it used for testing perhaps?)

File: drawers/cMouseDrawer.cpp [hunk 2]
```diff
@@ -167,21 +169,19 @@ void cMouseDrawer::drawToolTipBackground() {
...
-	assert(textWriter);
 	if (theStructure->belongsTo(player)) {
-		textWriter->writeWithOneInteger("Sight : %d", theStructure->getSight());
-		textWriter->writeWithOneInteger("Range : %d", theStructure->getRange());
+		textWriter.writeWithOneInteger("Sight : %d", theStructure->getSight());
+		textWriter.writeWithOneInteger("Range : %d", theStructure->getRange());
 	} else {
-		textWriter->write("Sight : Unknown");
-		textWriter->write("Range : Unknown");
+		textWriter.write("Range : Unknown");
```
Comment (stefanhendriks) [inline drawers/cMouseDrawer.cpp:178]: Range/Side order got swapped here. (compared to diff & compared to other block in `if`

File: drawers/cMouseDrawer.cpp [hunk 3]
```diff
@@ -190,58 +190,58 @@ void cMouseDrawer::drawToolTipGeneralInformation(cAbstractStructure * theStructu
 	} else {
 		sprintf(description, "%s", structureType.name);
 	}
-	textWriter->write(description, makecol(255, 255, 0));
-	textWriter->writeWithTwoIntegers("Hitpoints : %d/%d", theStructure->getHitPoints(), theStructure->getMaxHP());
-	textWriter->writeWithOneInteger("Armor : %d", theStructure->getArmor());
-	textWriter->writeWithOneInteger("Protected : %d%%", (100-theStructure->getPercentageNotPaved()));
+	textWriter.write(description, makecol(255, 255, 0));
+	textWriter.writeWithTwoIntegers("Hitpoints : %d/%d", theStructure->getHitPoints(), theStructure->getMaxHP());
```
Comment (stefanhendriks) [inline drawers/cMouseDrawer.cpp:194]: we can - eventually - get rid of `writeWith...` by combining `fmt::format` with `write()`. Not for now though :)

File: drawers/cMouseDrawer.cpp [hunk 4]
```diff
@@ -190,58 +190,58 @@ void cMouseDrawer::drawToolTipGeneralInformation(cAbstractStructure * theStructu
...
-	assert(textWriter);
-
+	
 	// TODO / IDEA --> Perhaps some 'spy on enemy intel' upgrade should be available for
 	// house Ordos, so you can actually check on the enemy spice etc.
 	if (theStructure->getOwner() == HUMAN) {
 		int spiceCapacityOfStructure = 0;
 		if (theStructure->getType() == REFINERY) {
 			cRefinery * refinery = dynamic_cast<cRefinery*>(theStructure);
+            assert(refinery);
```
Comment (stefanhendriks) [inline drawers/cMouseDrawer.cpp:226]: Additional asserts added? 

I do wonder if we can't solve this prettier. Instead of relying on `type` (REFINERY for instance), we should perhaps have some other way to get this information. Not sure how. My gut feel tells me that a sort of `composit` would make it possible to add some behavior to some structures which have a `capacity` of some sort. But I didn't spend too many brain cycles on this. Do you have any ideas about this?

File: drawers/cUnitDrawer.cpp [hunk 5]
```diff
@@ -1,27 +1,22 @@
...
 cUnitDrawer::cUnitDrawer() {
 }
 
-cUnitDrawer *cUnitDrawer::getInstance() {
-	if (instance == NULL) {
-		instance = new cUnitDrawer();
-	}
-
-	return instance;
+const cUnitDrawer& cUnitDrawer::getInstance() {
```
Comment (stefanhendriks) [inline drawers/cUnitDrawer.cpp:6]: afaik we don't use this drawer?

**Review by stefanhendriks — APPROVED** [review]
looks like a nice improvement to me, I've got a few minor questions.

I also want to test this a bit before merging. Seems to have quite some impact, but overall looking nice. Good to see more and more consistency in the code 🚀

File: drawers/cBuildingListDrawer.cpp [hunk 6]
```diff
@@ -1,19 +1,18 @@
...
+#include "cAllegroDrawer.h"
+#include "d2tmc.h"
+#include "data/gfxinter.h"
+#include "player/cPlayer.h"
 
 #include <allegro.h>
 
 cBuildingListDrawer::cBuildingListDrawer(cPlayer *player) :
-    m_textDrawer(new cTextDrawer(game_font)),
+    m_textDrawer(game_font),
```
[the remark this answers is not in the document]
Comment (pbomta) [reply drawers/cBuildingListDrawer.cpp:11]: Uhm, not really. This constructs `m_textDrawer` in-place inside `cBuildingListDrawer` so-to-speak. It's object memory is layout inside the object memory of `cBuildingListDrawer`. What you typed uses the copy constructor (unless that is optimized out - never know what the optimizer can do nowadays). So that would construct `cTextDrawer` on the stack and then copy (or move if it can) it into `cBuildingListDrawer`.

TL;DR: this is a way to use data member with non-default constructors.

File: drawers/cUnitDrawer.cpp (hunk 5 above)
Comment (stefanhendriks-felyx) [reply drawers/cUnitDrawer.cpp:6]: deleting is ok :)

File: drawers/cMouseDrawer.cpp (hunk 4 above)
Comment (stefanhendriks) [reply drawers/cMouseDrawer.cpp:226]: Yeah I can imagine. It feels awkward because we only have 1 or 2 buildings which will actually implement this :/
